### PR TITLE
feat(s3): add `allowedActionPatterns` parameter to grantWrite

### LINKED ISSUE #24074: iam: add more granular grant permissions
### Describe the feature

The IAM grant mechanism (grantWrite, grantRead, etc) should allow for additional parameters that can override the default permissions.

### Use Case

The default permissions granted using grantWrite, grantRead etc can be overly permissive for certain usecases.  For example `bucket.grantWrite(principale)` leads to the IAM permissions:

```
s3:Abort*
s3:DeleteObject*
s3:PutObject
s3:PutObjectLegalHold
s3:PutObjectRetention
s3:PutObjectTagging
s3:PutObjectVersionTagging
```

For certain usecases, these are overly permissive.  One example is granting a principal access to a write only bucket containing an audit log; the principal shouldn't be able to delete objects.

### Proposed Solution

One possible implementation:

```
const user = new iam.User(stack, 'audit-log-writer'))
const bucket = new s3.Bucket(this, 'audit-log-bucket', {...});
bucket.grantWrite(
  user, 
  null,
  ['s3:Abort*', 's3:DeleteObject*'] // proposed third option, which would include default actions to remove from the grant
)
```

### Other Information

_No response_

### Acknowledgements

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.63.2

### Environment details (OS name and version, etc.)

mac os

### PR DESCRIPTION
**Pull request by lpizzinidev**
I've added an optional parameter `allowedActionPatterns` that will restrict the permissions to a certain list of action patterns for the `grantWrite` method.

Closes #24074 

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by lpizzinidev** [comment]
> The pull request linter fails with the following errors:
> 
> ```
> ❌ Features must contain a change to a README file.
> ```
> 
> PRs must pass status checks before we can provide a meaningful review.
> 
> If you would like to request an exemption from the status checks or clarification on feedback, please leave a comment on this PR containing `Exemption Request` and/or `Clarification Request`.

Exemption Request.

The [S3 README](https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/aws-s3/README.md) doesn't contain references to the `grantWrite` method so there is nothing to change there.
Let me know if you'd like me to add an explanation of the method.

**Review by TheRealAmazonKendra — DISMISSED** [review]
Adding an exclusion filter doesn't really fit the pattern of any of the rest of these functions. I would prefer to see a function that explicitly adds the patterns a user wants than one that filters out patterns.

**Comment by lpizzinidev** [comment]
@TheRealAmazonKendra 
Thanks for the feedback!
I've changed the parameter to `allowedActions`, granting permission only to specified actions.
Please let me know if this is ok or if you need more changes.

File: packages/@aws-cdk/aws-s3/lib/bucket.ts [hunk 1]
```diff
@@ -204,8 +204,9 @@ export interface IBucket extends IResource {
    *
    * @param identity The principal
    * @param objectsKeyPattern Restrict the permission to a certain key pattern (default '*')
+   * @param allowedActions: Include only certain actions in the permissions
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-s3/lib/bucket.ts:207]:
```suggestion
   * @param allowedActionPatterns: Restrict the permissions to certain list of action patterns (default '*')
```

**Review by TheRealAmazonKendra — CHANGES_REQUESTED** [review]
Just a few suggested changes to the contract here.

**Comment by lpizzinidev** [comment]
@TheRealAmazonKendra 
Just a note, I removed the default value from the rest parameter since TS doesn't allow it and handled the default to `this.writeAction` in the function logic.

File: packages/@aws-cdk/aws-s3/lib/bucket.ts [hunk 2]
```diff
@@ -204,8 +204,9 @@ export interface IBucket extends IResource {
...
    */
-  grantWrite(identity: iam.IGrantable, objectsKeyPattern?: any): iam.Grant;
+  grantWrite(identity: iam.IGrantable, objectsKeyPattern?: any, allowedActions?: string[]): iam.Grant;
```
[the remark this answers is not in the document]
Comment (TheRealAmazonKendra) [reply packages/@aws-cdk/aws-s3/lib/bucket.ts]: Ah, that's right. Fine as is.

**Comment by TheRealAmazonKendra** [comment]
Go ahead and change that back to an optional array. It becomes a breaking change if the field is required.

**Review by TheRealAmazonKendra — APPROVED** [review]

**Pull request merged**